src/main/java/io/vertx/core/http/impl/WebSocketImplBase.java

```diff
@@ -195,73 +219,116 @@ void handleFrame(WebSocketFrameInternal frame) {
       if (frameHandler != null) {
         frameHandler.handle(frame);
       }
+    }
+  }
 
-      if (textMessageHandler != null && (frame.type() == FrameType.TEXT
-        || (processingPartialTextMessage && frame.type() == FrameType.CONTINUATION))) {
-        handleTextFrame(frame);
+  private class FrameAggregator implements Handler<WebSocketFrameInternal> {
+
+    private Handler<String> textMessageHandler;
+    private Handler<Buffer> binaryMessageHandler;
+    private Buffer textMessageBuffer;
+    private Buffer binaryMessageBuffer;
+
+    @Override
+    public void handle(WebSocketFrameInternal frame) {
+      switch (frame.type()) {
+        case TEXT:
+          handleTextFrame(frame);
+          break;
+        case BINARY:
+          handleBinaryFrame(frame);
+          break;
+        case CONTINUATION:
+          if (textMessageBuffer != null && textMessageBuffer.length() > 0) {
+            handleTextFrame(frame);
+          } else if (binaryMessageBuffer != null && binaryMessageBuffer.length() > 0) {
+            handleBinaryFrame(frame);
+          }
+          break;
       }
+    }
 
-      if (binaryMessageHandler != null && (frame.type() == FrameType.BINARY
-        || (processingPartialBinaryMessage && frame.type() == FrameType.CONTINUATION))) {
-        handleBinaryFrame(frame);
+    private void handleTextFrame(WebSocketFrameInternal frame) {
+      Buffer frameBuffer = Buffer.buffer(frame.getBinaryData());
+      if (textMessageBuffer == null) {
+        textMessageBuffer = frameBuffer;
```

Comment:
Any message that is split into at least 3 frames will have a continuation frame that is not final.

I see now that your change checks if buffer != null instead of having a separate processingPartialMessage flag.  That made it more difficult for me to see that you were correctly dropping continuation frames after the failure.  Now that I have looked through it again, I think everything is ok.